# Project - 1: Weather Dataset

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv("Weather Dataset.csv")

In [3]:
data

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0        1/1/2012 0:00                 -1.8              -3.9         86   
1        1/1/2012 1:00                 -1.8              -3.7         87   
2        1/1/2012 2:00                 -1.8              -3.4         89   
3        1/1/2012 3:00                 -1.5              -3.2         88   
4        1/1/2012 4:00                 -1.5              -3.3         88   
...                ...                  ...               ...        ...   
8779  12/31/2012 19:00                  0.1              -2.7         81   
8780  12/31/2012 20:00                  0.2              -2.4         83   
8781  12/31/2012 21:00                 -0.5              -1.5         93   
8782  12/31/2012 22:00                 -0.2              -1.8         89   
8783  12/31/2012 23:00                  0.0              -2.1         86   

      Wind Speed_km/h  Visibility_km  Press_kPa               Weather  
0                   4            8.0     101.24                   Fog  
1                   4            8.0     101.24                   Fog  
2                   7            4.0     101.26  Freezing Drizzle,Fog  
3                   6            4.0     101.27  Freezing Drizzle,Fog  
4                   7            4.8     101.23                   Fog  
...               ...            ...        ...                   ...  
8779               30            9.7     100.13                  Snow  
8780               24            9.7     100.03                  Snow  
8781               28            4.8      99.95                  Snow  
8782               28            9.7      99.91                  Snow  
8783               30           11.3      99.89                  Snow  

[8784 rows x 8 columns]

# Analyzing Dataframes

In [5]:
data.tail(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
8783  12/31/2012 23:00                  0.0              -2.1         86   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather  
8783               30           11.3      99.89    Snow

In [5]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather  
0                4            8.0     101.24     Fog

In [8]:
data.shape

(8784, 8)

In [9]:
data.index

RangeIndex(start=0, stop=8784, step=1)

In [10]:
data.columns

Index(['Date/Time', 'Temp_degree celcius', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

In [11]:
data.dtypes

Date/Time               object
Temp_degree celcius    float64
Dew Point Temp_C       float64
Rel Hum_%                int64
Wind Speed_km/h          int64
Visibility_km          float64
Press_kPa              float64
Weather                 object
dtype: object

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Date/Time            8784 non-null   object 
 1   Temp_degree celcius  8784 non-null   float64
 2   Dew Point Temp_C     8784 non-null   float64
 3   Rel Hum_%            8784 non-null   int64  
 4   Wind Speed_km/h      8784 non-null   int64  
 5   Visibility_km        8784 non-null   float64
 6   Press_kPa            8784 non-null   float64
 7   Weather              8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


In [17]:
data["Rel Hum_%"].unique()

array([ 86,  87,  89,  88,  85,  91,  82,  81,  84,  79,  77,  72,  71,
        69,  70,  68,  67,  66,  56,  57,  59,  55,  54,  58,  65,  53,
        60,  61,  63,  52,  50,  49,  62,  51,  75,  73,  80,  83,  76,
        64,  74,  78,  47,  90,  48,  45,  46,  92,  93,  42,  43,  40,
        39,  44,  94,  37,  41,  38,  96,  98,  99,  97,  95, 100,  29,
        28,  27,  19,  20,  18,  25,  26,  30,  24,  34,  33,  31,  32,
        35,  36,  21,  22,  23])

In [18]:
data.nunique()

Date/Time              8784
Temp_degree celcius     533
Dew Point Temp_C        489
Rel Hum_%                83
Wind Speed_km/h          34
Visibility_km            24
Press_kPa               518
Weather                  50
dtype: int64

In [7]:
data["Rel Hum_%"].nunique()

83

In [19]:
data.count()

Date/Time              8784
Temp_degree celcius    8784
Dew Point Temp_C       8784
Rel Hum_%              8784
Wind Speed_km/h        8784
Visibility_km          8784
Press_kPa              8784
Weather                8784
dtype: int64

In [26]:
data["Wind Speed_km/h"].count()

np.int64(8784)

In [21]:
data["Wind Speed_km/h"].value_counts()

Wind Speed_km/h
9     830
11    791
13    735
15    719
7     677
17    666
19    616
6     609
20    496
4     474
22    439
24    374
0     309
26    242
28    205
30    161
32    139
33     85
35     53
37     45
39     24
41     22
44     14
43     13
48     13
46     11
52      7
57      5
50      4
2       2
70      1
83      1
63      1
54      1
Name: count, dtype: int64

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Date/Time            8784 non-null   object 
 1   Temp_degree celcius  8784 non-null   float64
 2   Dew Point Temp_C     8784 non-null   float64
 3   Rel Hum_%            8784 non-null   int64  
 4   Wind Speed_km/h      8784 non-null   int64  
 5   Visibility_km        8784 non-null   float64
 6   Press_kPa            8784 non-null   float64
 7   Weather              8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


# Q.1. Find all the unique "Wind Speed" values in the data.

In [34]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather  
0                4            8.0     101.24     Fog

In [27]:
data.nunique()

Date/Time              8784
Temp_degree celcius     533
Dew Point Temp_C        489
Rel Hum_%                83
Wind Speed_km/h          34
Visibility_km            24
Press_kPa               518
Weather                  50
dtype: int64

* Number of total 'Wind Speed_km/h'

In [32]:
data["Wind Speed_km/h"].nunique()

34

* Total 'Wind Speed_km/h' values

In [33]:
data["Wind Speed_km/h"].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

# Q.2. Find the number of times when the 'Weather is exactly clear'.

In [35]:
# For determining the number of times, we need to use .value_counts() function.

data["Weather"].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

So, the answer to the number of times the weather is exactly clear is 1326 times.

* Alternative Way

In [41]:
# Filtering data
data[data.Weather == 'Clear']

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
67      1/3/2012 19:00                -16.9             -24.8         50   
114     1/5/2012 18:00                 -7.1             -14.4         56   
115     1/5/2012 19:00                 -9.2             -15.4         61   
116     1/5/2012 20:00                 -9.8             -15.7         62   
117     1/5/2012 21:00                 -9.0             -14.8         63   
...                ...                  ...               ...        ...   
8646   12/26/2012 6:00                -13.4             -14.8         89   
8698  12/28/2012 10:00                 -6.1              -8.6         82   
8713   12/29/2012 1:00                -11.9             -13.6         87   
8714   12/29/2012 2:00                -11.8             -13.1         90   
8756  12/30/2012 20:00                -13.8             -16.5         80   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather  
67                 24           25.0     101.74   Clear  
114                11           25.0     100.71   Clear  
115                 7           25.0     100.80   Clear  
116                 9           25.0     100.83   Clear  
117                13           25.0     100.83   Clear  
...               ...            ...        ...     ...  
8646                4           25.0     102.47   Clear  
8698               19           24.1     101.27   Clear  
8713               11           25.0     101.31   Clear  
8714               13           25.0     101.33   Clear  
8756               24           25.0     101.52   Clear  

[1326 rows x 8 columns]

In [44]:
# groupby
data.groupby('Weather').get_group('Clear')

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
67      1/3/2012 19:00                -16.9             -24.8         50   
114     1/5/2012 18:00                 -7.1             -14.4         56   
115     1/5/2012 19:00                 -9.2             -15.4         61   
116     1/5/2012 20:00                 -9.8             -15.7         62   
117     1/5/2012 21:00                 -9.0             -14.8         63   
...                ...                  ...               ...        ...   
8646   12/26/2012 6:00                -13.4             -14.8         89   
8698  12/28/2012 10:00                 -6.1              -8.6         82   
8713   12/29/2012 1:00                -11.9             -13.6         87   
8714   12/29/2012 2:00                -11.8             -13.1         90   
8756  12/30/2012 20:00                -13.8             -16.5         80   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather  
67                 24           25.0     101.74   Clear  
114                11           25.0     100.71   Clear  
115                 7           25.0     100.80   Clear  
116                 9           25.0     100.83   Clear  
117                13           25.0     100.83   Clear  
...               ...            ...        ...     ...  
8646                4           25.0     102.47   Clear  
8698               19           24.1     101.27   Clear  
8713               11           25.0     101.31   Clear  
8714               13           25.0     101.33   Clear  
8756               24           25.0     101.52   Clear  

[1326 rows x 8 columns]

In [47]:
data.Weather.value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

# Q.3. Find the number of times when the Wind Speed is exactly 4 km/h.

In [53]:
data["Wind Speed_km/h"].value_counts()

Wind Speed_km/h
9     830
11    791
13    735
15    719
7     677
17    666
19    616
6     609
20    496
4     474
22    439
24    374
0     309
26    242
28    205
30    161
32    139
33     85
35     53
37     45
39     24
41     22
44     14
43     13
48     13
46     11
52      7
57      5
50      4
2       2
70      1
83      1
63      1
54      1
Name: count, dtype: int64

In [57]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather  
0                4            8.0     101.24     Fog

In [58]:
data.groupby("Wind Speed_km/h").get_group(4)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0        1/1/2012 0:00                 -1.8              -3.9         86   
1        1/1/2012 1:00                 -1.8              -3.7         87   
96       1/5/2012 0:00                 -8.8             -11.7         79   
101      1/5/2012 5:00                 -7.0              -9.5         82   
146      1/7/2012 2:00                 -8.1             -11.1         79   
...                ...                  ...               ...        ...   
8768   12/31/2012 8:00                 -8.6             -10.3         87   
8769   12/31/2012 9:00                 -8.1              -9.6         89   
8770  12/31/2012 10:00                 -7.4              -8.9         89   
8772  12/31/2012 12:00                 -5.8              -7.5         88   
8773  12/31/2012 13:00                 -4.6              -6.6         86   

      Wind Speed_km/h  Visibility_km  Press_kPa       Weather  
0                   4            8.0     101.24           Fog  
1                   4            8.0     101.24           Fog  
96                  4            9.7     100.32          Snow  
101                 4            4.0     100.19          Snow  
146                 4           19.3     100.15        Cloudy  
...               ...            ...        ...           ...  
8768                4            3.2     101.14  Snow Showers  
8769                4            2.4     101.09          Snow  
8770                4            6.4     101.05      Snow,Fog  
8772                4           12.9     100.78          Snow  
8773                4           12.9     100.63          Snow  

[474 rows x 8 columns]

* IMPORTANT: When using get_group() function, if we want to find out how many times a number is in a column, we don't need to put that number in double quotes. 
             But, if we use a string in the parenthesis, we must use double quotes ("str").

In [13]:
data[data["Wind Speed_km/h"] == 4]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0        1/1/2012 0:00                 -1.8              -3.9         86   
1        1/1/2012 1:00                 -1.8              -3.7         87   
96       1/5/2012 0:00                 -8.8             -11.7         79   
101      1/5/2012 5:00                 -7.0              -9.5         82   
146      1/7/2012 2:00                 -8.1             -11.1         79   
...                ...                  ...               ...        ...   
8768   12/31/2012 8:00                 -8.6             -10.3         87   
8769   12/31/2012 9:00                 -8.1              -9.6         89   
8770  12/31/2012 10:00                 -7.4              -8.9         89   
8772  12/31/2012 12:00                 -5.8              -7.5         88   
8773  12/31/2012 13:00                 -4.6              -6.6         86   

      Wind Speed_km/h  Visibility_km  Press_kPa       Weather  
0                   4            8.0     101.24           Fog  
1                   4            8.0     101.24           Fog  
96                  4            9.7     100.32          Snow  
101                 4            4.0     100.19          Snow  
146                 4           19.3     100.15        Cloudy  
...               ...            ...        ...           ...  
8768                4            3.2     101.14  Snow Showers  
8769                4            2.4     101.09          Snow  
8770                4            6.4     101.05      Snow,Fog  
8772                4           12.9     100.78          Snow  
8773                4           12.9     100.63          Snow  

[474 rows x 8 columns]

In [14]:
# The answer is 474

# Q.4. Find out all the null values in the data.

In [16]:
data.isnull().sum()

Date/Time              0
Temp_degree celcius    0
Dew Point Temp_C       0
Rel Hum_%              0
Wind Speed_km/h        0
Visibility_km          0
Press_kPa              0
Weather                0
dtype: int64

In [17]:
# Answer = no null values in any column.

In [18]:
data.notnull().sum()

Date/Time              8784
Temp_degree celcius    8784
Dew Point Temp_C       8784
Rel Hum_%              8784
Wind Speed_km/h        8784
Visibility_km          8784
Press_kPa              8784
Weather                8784
dtype: int64

In [19]:
# Every column is not null.

# Q.5. Rename the column name 'Weather' of the dataframe to 'Weather Condition'.

In [4]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather  
0                4            8.0     101.24     Fog

In [3]:
data.rename(columns = {'Weather' : 'Weather Condition'}, inplace = True)

In [4]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
0                4            8.0     101.24               Fog

In [8]:
# To rename a column's name permanently, we must include "inplace = True" in the rename function. Otherwise it will act as a temporary rename, and will go back to the previous name afterward.

# Q.6. What is the mean 'Visibility'?

In [9]:
data.Visibility_km.mean()

np.float64(27.664446721311478)

# Q.7. What is the standard deviation of 'Pressure' in this data?

In [10]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
0                4            8.0     101.24               Fog

In [11]:
data.Press_kPa.std()

np.float64(0.8440047459486483)

# Q.8. What is the variance of 'Relative Humidity' in this data?

In [4]:
data['Rel Hum_%'].var()

np.float64(286.24855019850196)

# Q.9. Find all instances when "Snow" was recorded.

In [13]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
0                4            8.0     101.24               Fog

* Way 1: groupby

In [18]:
data.groupby("Weather Condition").get_group("Snow")

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
55       1/3/2012 7:00                -14.0             -19.5         63   
84      1/4/2012 12:00                -13.7             -21.7         51   
86      1/4/2012 14:00                -11.3             -19.0         53   
87      1/4/2012 15:00                -10.2             -16.3         61   
88      1/4/2012 16:00                 -9.4             -15.5         61   
...                ...                  ...               ...        ...   
8779  12/31/2012 19:00                  0.1              -2.7         81   
8780  12/31/2012 20:00                  0.2              -2.4         83   
8781  12/31/2012 21:00                 -0.5              -1.5         93   
8782  12/31/2012 22:00                 -0.2              -1.8         89   
8783  12/31/2012 23:00                  0.0              -2.1         86   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
55                 19           25.0     100.95              Snow  
84                 11           24.1     101.25              Snow  
86                  7           19.3     100.97              Snow  
87                 11            9.7     100.89              Snow  
88                 13           19.3     100.79              Snow  
...               ...            ...        ...               ...  
8779               30            9.7     100.13              Snow  
8780               24            9.7     100.03              Snow  
8781               28            4.8      99.95              Snow  
8782               28            9.7      99.91              Snow  
8783               30           11.3      99.89              Snow  

[390 rows x 8 columns]

In [15]:
# Snow was recorded 390 times.

* Way 2: value_counts

In [19]:
data["Weather Condition"].value_counts()

Weather Condition
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle      

In [20]:
# Answer = 390 times

* Way 3: filtering

In [22]:
data[data["Weather Condition"] == "Snow"]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
55       1/3/2012 7:00                -14.0             -19.5         63   
84      1/4/2012 12:00                -13.7             -21.7         51   
86      1/4/2012 14:00                -11.3             -19.0         53   
87      1/4/2012 15:00                -10.2             -16.3         61   
88      1/4/2012 16:00                 -9.4             -15.5         61   
...                ...                  ...               ...        ...   
8779  12/31/2012 19:00                  0.1              -2.7         81   
8780  12/31/2012 20:00                  0.2              -2.4         83   
8781  12/31/2012 21:00                 -0.5              -1.5         93   
8782  12/31/2012 22:00                 -0.2              -1.8         89   
8783  12/31/2012 23:00                  0.0              -2.1         86   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
55                 19           25.0     100.95              Snow  
84                 11           24.1     101.25              Snow  
86                  7           19.3     100.97              Snow  
87                 11            9.7     100.89              Snow  
88                 13           19.3     100.79              Snow  
...               ...            ...        ...               ...  
8779               30            9.7     100.13              Snow  
8780               24            9.7     100.03              Snow  
8781               28            4.8      99.95              Snow  
8782               28            9.7      99.91              Snow  
8783               30           11.3      99.89              Snow  

[390 rows x 8 columns]

In [1]:
# Answer 390 times.

* Way 4: str.contains

In [5]:
data[data["Weather Condition"].str.contains("Snow")]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
41      1/2/2012 17:00                 -2.1              -9.5         57   
44      1/2/2012 20:00                 -5.6             -13.4         54   
45      1/2/2012 21:00                 -5.8             -12.8         58   
47      1/2/2012 23:00                 -7.4             -14.1         59   
48       1/3/2012 0:00                 -9.0             -16.0         57   
...                ...                  ...               ...        ...   
8779  12/31/2012 19:00                  0.1              -2.7         81   
8780  12/31/2012 20:00                  0.2              -2.4         83   
8781  12/31/2012 21:00                 -0.5              -1.5         93   
8782  12/31/2012 22:00                 -0.2              -1.8         89   
8783  12/31/2012 23:00                  0.0              -2.1         86   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
41                 22           25.0      99.66      Snow Showers  
44                 24           25.0     100.07      Snow Showers  
45                 26           25.0     100.15      Snow Showers  
47                 17           19.3     100.27      Snow Showers  
48                 28           25.0     100.35      Snow Showers  
...               ...            ...        ...               ...  
8779               30            9.7     100.13              Snow  
8780               24            9.7     100.03              Snow  
8781               28            4.8      99.95              Snow  
8782               28            9.7      99.91              Snow  
8783               30           11.3      99.89              Snow  

[583 rows x 8 columns]

In [6]:
# 583 times where snow is fully or partially found.

In [9]:
data[data["Weather Condition"].str.contains("Snow")].head(50)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
41   1/2/2012 17:00                 -2.1              -9.5         57   
44   1/2/2012 20:00                 -5.6             -13.4         54   
45   1/2/2012 21:00                 -5.8             -12.8         58   
47   1/2/2012 23:00                 -7.4             -14.1         59   
48    1/3/2012 0:00                 -9.0             -16.0         57   
50    1/3/2012 2:00                -10.5             -15.8         65   
51    1/3/2012 3:00                -11.3             -18.7         54   
53    1/3/2012 5:00                -12.9             -19.1         60   
54    1/3/2012 6:00                -13.3             -19.3         61   
55    1/3/2012 7:00                -14.0             -19.5         63   
84   1/4/2012 12:00                -13.7             -21.7         51   
86   1/4/2012 14:00                -11.3             -19.0         53   
87   1/4/2012 15:00                -10.2             -16.3         61   
88   1/4/2012 16:00                 -9.4             -15.5         61   
89   1/4/2012 17:00                 -8.9             -13.2         71   
90   1/4/2012 18:00                 -8.9             -12.6         75   
91   1/4/2012 19:00                 -8.4             -12.7         71   
92   1/4/2012 20:00                 -7.8             -12.1         71   
93   1/4/2012 21:00                 -7.6             -11.6         73   
94   1/4/2012 22:00                 -9.5             -12.7         77   
95   1/4/2012 23:00                 -9.6             -12.6         79   
96    1/5/2012 0:00                 -8.8             -11.7         79   
97    1/5/2012 1:00                 -7.5             -10.2         81   
98    1/5/2012 2:00                 -5.4              -8.3         80   
99    1/5/2012 3:00                 -5.0              -7.7         81   
100   1/5/2012 4:00                 -4.1              -6.5         83   
101   1/5/2012 5:00                 -7.0              -9.5         82   
102   1/5/2012 6:00                 -8.2             -10.7         82   
103   1/5/2012 7:00                 -7.1              -9.7         82   
104   1/5/2012 8:00                 -6.1              -9.1         79   
123   1/6/2012 3:00                -10.6             -16.0         64   
124   1/6/2012 4:00                -11.3             -16.1         68   
125   1/6/2012 5:00                -11.8             -16.0         71   
126   1/6/2012 6:00                -12.0             -16.2         71   
127   1/6/2012 7:00                -14.4             -16.3         85   
128   1/6/2012 8:00                -12.3             -16.2         73   
129   1/6/2012 9:00                -12.5             -16.7         71   
130  1/6/2012 10:00                -12.3             -16.3         72   
131  1/6/2012 11:00                -12.0             -16.0         72   
132  1/6/2012 12:00                -11.7             -15.4         74   
133  1/6/2012 13:00                -11.9             -15.6         74   
134  1/6/2012 14:00                -11.2             -14.8         75   
135  1/6/2012 15:00                -11.5             -14.4         79   
136  1/6/2012 16:00                -11.6             -14.7         78   
137  1/6/2012 17:00                -11.2             -14.3         78   
158  1/7/2012 14:00                 -4.4              -6.9         83   
159  1/7/2012 15:00                 -3.7              -6.1         83   
160  1/7/2012 16:00                 -3.6              -5.9         84   
161  1/7/2012 17:00                 -3.1              -5.4         84   
162  1/7/2012 18:00                 -3.2              -5.3         85   

     Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
41                22           25.0      99.66      Snow Showers  
44                24           25.0     100.07      Snow Showers  
45                26           25.0     100.15      Snow Showers  
47                

In [9]:
# Snow found fully or partially in the first 50 weather data.

In [8]:
data[data["Weather Condition"].str.contains("Snow")].tail(50)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
8680  12/27/2012 16:00                 -4.5              -6.2         88   
8681  12/27/2012 17:00                 -4.2              -5.9         88   
8682  12/27/2012 18:00                 -4.0              -5.7         88   
8683  12/27/2012 19:00                 -3.9              -5.6         88   
8684  12/27/2012 20:00                 -3.7              -5.3         89   
8685  12/27/2012 21:00                 -3.7              -4.8         92   
8686  12/27/2012 22:00                 -3.8              -4.6         94   
8687  12/27/2012 23:00                 -4.0              -5.6         89   
8688   12/28/2012 0:00                 -4.2              -5.7         89   
8689   12/28/2012 1:00                 -4.4              -6.6         85   
8690   12/28/2012 2:00                 -4.3              -6.3         86   
8691   12/28/2012 3:00                 -4.6              -5.9         91   
8692   12/28/2012 4:00                 -4.9              -5.9         93   
8723  12/29/2012 11:00                -10.9             -12.2         90   
8724  12/29/2012 12:00                -10.5             -11.6         92   
8725  12/29/2012 13:00                -10.0             -11.1         92   
8726  12/29/2012 14:00                 -9.3             -10.5         91   
8727  12/29/2012 15:00                 -8.8             -10.0         91   
8728  12/29/2012 16:00                 -8.5              -9.9         90   
8729  12/29/2012 17:00                 -9.0             -10.4         90   
8730  12/29/2012 18:00                 -9.3             -10.9         88   
8731  12/29/2012 19:00                 -9.5             -11.2         87   
8732  12/29/2012 20:00                 -9.7             -11.6         86   
8733  12/29/2012 21:00                 -9.8             -11.8         85   
8734  12/29/2012 22:00                -10.1             -11.6         89   
8735  12/29/2012 23:00                -10.0             -12.0         85   
8736   12/30/2012 0:00                 -9.6             -11.3         87   
8737   12/30/2012 1:00                 -9.4             -10.5         92   
8738   12/30/2012 2:00                 -9.3             -10.4         92   
8739   12/30/2012 3:00                 -9.1             -10.4         90   
8740   12/30/2012 4:00                 -9.3             -10.6         90   
8741   12/30/2012 5:00                 -9.1             -10.4         90   
8742   12/30/2012 6:00                 -9.3             -10.8         89   
8767   12/31/2012 7:00                 -9.3             -11.3         85   
8768   12/31/2012 8:00                 -8.6             -10.3         87   
8769   12/31/2012 9:00                 -8.1              -9.6         89   
8770  12/31/2012 10:00                 -7.4              -8.9         89   
8771  12/31/2012 11:00                 -6.7              -7.9         91   
8772  12/31/2012 12:00                 -5.8              -7.5         88   
8773  12/31/2012 13:00                 -4.6              -6.6         86   
8774  12/31/2012 14:00                 -3.4              -5.7         84   
8775  12/31/2012 15:00                 -2.3              -4.6         84   
8776  12/31/2012 16:00                 -1.4              -4.0         82   
8777  12/31/2012 17:00                 -1.1              -3.3         85   
8778  12/31/2012 18:00                 -1.3              -3.1         88   
8779  12/31/2012 19:00                  0.1              -2.7         81   
8780  12/31/2012 20:00                  0.2              -2.4         83   
8781  12/31/2012 21:00                 -0.5              -1.5         93   
8782  12/31/2012 22:00                 -0.2              -1.8         89   
8783  12/31/2012 23:00                  0.0              -2.1         86   

      Wind Speed_km/h  Visibility_km  Press_kPa      Weather Condition  
8680               37            2.0     100.44      Snow,Blowi

In [10]:
# Snow found fully or partially in the last 50 weather data.

# Q.10. Find all the instances when the 'Wind Speed_km/h' is above 24 and 'Visibility' is 25.

In [13]:
data[(data['Wind Speed_km/h'] > 24) & (data['Visibility_km'] == 25)]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
23      1/1/2012 23:00                  5.3               2.0         79   
24       1/2/2012 0:00                  5.2               1.5         77   
25       1/2/2012 1:00                  4.6               0.0         72   
26       1/2/2012 2:00                  3.9              -0.9         71   
27       1/2/2012 3:00                  3.7              -1.5         69   
...                ...                  ...               ...        ...   
8705  12/28/2012 17:00                 -8.6             -12.0         76   
8753  12/30/2012 17:00                -12.1             -15.8         74   
8755  12/30/2012 19:00                -13.4             -16.5         77   
8759  12/30/2012 23:00                -12.1             -15.1         78   
8760   12/31/2012 0:00                -11.1             -14.4         77   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
23                 30           25.0      99.31            Cloudy  
24                 35           25.0      99.26      Rain Showers  
25                 39           25.0      99.26            Cloudy  
26                 32           25.0      99.26     Mostly Cloudy  
27                 33           25.0      99.30     Mostly Cloudy  
...               ...            ...        ...               ...  
8705               26           25.0     101.34      Mainly Clear  
8753               28           25.0     101.26      Mainly Clear  
8755               26           25.0     101.47      Mainly Clear  
8759               28           25.0     101.52     Mostly Cloudy  
8760               26           25.0     101.51            Cloudy  

[308 rows x 8 columns]

In [19]:
# Answer = 308 times.

# Q.11. What is the Mean value of each column against each 'Weather Condition'?

In [9]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
0                4            8.0     101.24               Fog

In [11]:
data.groupby("Weather Condition").mean(numeric_only=True)

Temp_degree celcius  \
Weather Condition                                              
Clear                                               6.825716   
Cloudy                                              7.970544   
Drizzle                                             7.353659   
Drizzle,Fog                                         8.067500   
Drizzle,Ice Pellets,Fog                             0.400000   
Drizzle,Snow                                        1.050000   
Drizzle,Snow,Fog                                    0.693333   
Fog                                                 4.303333   
Freezing Drizzle                                   -5.657143   
Freezing Drizzle,Fog                               -2.533333   
Freezing Drizzle,Haze                              -5.433333   
Freezing Drizzle,Snow                              -5.109091   
Freezing Fog                                       -7.575000   
Freezing Rain                                      -3.885714   
Freezing Rain,Fog                                  -2.225000   
Freezing Rain,Haze                                 -4.900000   
Freezing Rain,Ice Pellets,Fog                      -2.600000   
Freezing Rain,Snow Grains                          -5.000000   
Haze                                               -0.200000   
Mainly Clear                                       12.558927   
Moderate Rain,Fog                                   1.700000   
Moderate Snow                                      -5.525000   
Moderate Snow,Blowing Snow                         -5.450000   
Mostly Cloudy                                      10.574287   
Rain                                                9.786275   
Rain Showers                                       13.722340   
Rain Showers,Fog                                   12.800000   
Rain Showers,Snow Showers                           2.150000   
Rain,Fog                                            8.273276   
Rain,Haze                                           4.633333   
Rain,Ice Pellets                                    0.600000   
Rain,Snow                                           1.055556   
Rain,Snow Grains                                    1.900000   
Rain,Snow,Fog                                       0.800000   
Rain,Snow,Ice Pellets                               1.100000   
Snow                                               -4.524103   
Snow Pellets                                        0.700000   
Snow Showers                                       -3.506667   
Snow Showers,Fog                                  -10.675000   
Snow,Blowing Snow                                  -5.410526   
Snow,Fog                                           -5.075676   
Snow,Haze                                          -4.020000   
Snow,Ice Pellets                                   -1.883333   
Thunderstorms                                      24.150000   
Thunderstorms,Heavy Rain Showers                   10.900000   
Thunderstorms,Moderate Rain Showers,Fog            19.600000   
Thunderstorms,Rain                                 20.433333   
Thunderstorms,Rain Showers                         20.037500   
Thunderstorms,Rain Showers,Fog                     21.600000   
Thunderstorms,Rain,Fog                             20.600000   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                            0.089367  64.497738   
Cloudy                                           2.375810  69.592593   
Drizzle                                          5.504878  88.243902   
Drizzle,Fog                                      7.033750  93.275000   
Drizzle,Ice Pellets,Fog                         -0.700000  92.000000   
Drizzle,Snow                                     0.150000  93.500000   
Drizzle,Snow,Fog                                 0.120000  95.866667   
Fog                                              3.159333  92.28

In [12]:
'''If we don't write (numeric only = True) in the bracket of mean, then it will show errors, because the Date/Time and Weather Condition columns are strings. That's why the mean() function will not
be able to calculate the mean of the whole dataset. 

To avoid this error, we must put the mean() function as mean(numeric_only = True).'''

"If we don't write (numeric only = True) in the bracket of mean, then it will show errors, because the Date/Time and Weather Condition columns are strings. That's why the mean() function will not\nbe able to calculate the mean of the whole dataset. \n\nTo avoid this error, we must put the mean() function as mean(numeric_only = True)."

# Q.12. What is the minimum and the maximum value of each column against the 'Weather Condition' column? 

In [5]:
data.head(1)

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0  1/1/2012 0:00                 -1.8              -3.9         86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
0                4            8.0     101.24               Fog

In [6]:
data.groupby('Weather Condition').min(numeric_only=True)

Temp_degree celcius  \
Weather Condition                                              
Clear                                                  -23.3   
Cloudy                                                 -21.4   
Drizzle                                                  1.1   
Drizzle,Fog                                              0.0   
Drizzle,Ice Pellets,Fog                                  0.4   
Drizzle,Snow                                             0.9   
Drizzle,Snow,Fog                                         0.3   
Fog                                                    -16.0   
Freezing Drizzle                                        -9.0   
Freezing Drizzle,Fog                                    -6.4   
Freezing Drizzle,Haze                                   -5.8   
Freezing Drizzle,Snow                                   -8.3   
Freezing Fog                                           -19.0   
Freezing Rain                                           -6.5   
Freezing Rain,Fog                                       -6.1   
Freezing Rain,Haze                                      -4.9   
Freezing Rain,Ice Pellets,Fog                           -2.6   
Freezing Rain,Snow Grains                               -5.0   
Haze                                                   -11.5   
Mainly Clear                                           -22.8   
Moderate Rain,Fog                                        1.7   
Moderate Snow                                           -6.3   
Moderate Snow,Blowing Snow                              -5.5   
Mostly Cloudy                                          -23.2   
Rain                                                     0.3   
Rain Showers                                             1.6   
Rain Showers,Fog                                        12.8   
Rain Showers,Snow Showers                                2.1   
Rain,Fog                                                 0.0   
Rain,Haze                                                4.0   
Rain,Ice Pellets                                         0.6   
Rain,Snow                                                0.6   
Rain,Snow Grains                                         1.9   
Rain,Snow,Fog                                            0.8   
Rain,Snow,Ice Pellets                                    0.9   
Snow                                                   -16.7   
Snow Pellets                                             0.7   
Snow Showers                                           -13.3   
Snow Showers,Fog                                       -11.3   
Snow,Blowing Snow                                      -12.0   
Snow,Fog                                               -10.1   
Snow,Haze                                               -4.3   
Snow,Ice Pellets                                        -4.3   
Thunderstorms                                           21.6   
Thunderstorms,Heavy Rain Showers                        10.9   
Thunderstorms,Moderate Rain Showers,Fog                 19.6   
Thunderstorms,Rain                                      19.4   
Thunderstorms,Rain Showers                              11.0   
Thunderstorms,Rain Showers,Fog                          19.5   
Thunderstorms,Rain,Fog                                  20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                               -28.5         20   
Cloudy                                              -26.8         18   
Drizzle                                              -0.2         74   
Drizzle,Fog                                          -1.6         85   
Drizzle,Ice Pellets,Fog                              -0.7         92   
Drizzle,Snow                                          0.1         92   
Drizzle,Snow,Fog                                     -0.1         92   
Fog                                                 -17.2       

In [8]:
data.groupby("Weather Condition").max(numeric_only=True)

Temp_degree celcius  \
Weather Condition                                              
Clear                                                   32.8   
Cloudy                                                  30.5   
Drizzle                                                 18.8   
Drizzle,Fog                                             19.9   
Drizzle,Ice Pellets,Fog                                  0.4   
Drizzle,Snow                                             1.2   
Drizzle,Snow,Fog                                         1.1   
Fog                                                     20.8   
Freezing Drizzle                                        -2.3   
Freezing Drizzle,Fog                                    -0.3   
Freezing Drizzle,Haze                                   -5.0   
Freezing Drizzle,Snow                                   -3.3   
Freezing Fog                                            -0.1   
Freezing Rain                                            0.3   
Freezing Rain,Fog                                        0.1   
Freezing Rain,Haze                                      -4.9   
Freezing Rain,Ice Pellets,Fog                           -2.6   
Freezing Rain,Snow Grains                               -5.0   
Haze                                                    14.1   
Mainly Clear                                            33.0   
Moderate Rain,Fog                                        1.7   
Moderate Snow                                           -4.9   
Moderate Snow,Blowing Snow                              -5.4   
Mostly Cloudy                                           32.4   
Rain                                                    22.8   
Rain Showers                                            26.4   
Rain Showers,Fog                                        12.8   
Rain Showers,Snow Showers                                2.2   
Rain,Fog                                                21.7   
Rain,Haze                                                5.5   
Rain,Ice Pellets                                         0.6   
Rain,Snow                                                1.7   
Rain,Snow Grains                                         1.9   
Rain,Snow,Fog                                            0.8   
Rain,Snow,Ice Pellets                                    1.3   
Snow                                                     3.7   
Snow Pellets                                             0.7   
Snow Showers                                             2.9   
Snow Showers,Fog                                       -10.0   
Snow,Blowing Snow                                       -1.4   
Snow,Fog                                                 1.1   
Snow,Haze                                               -3.6   
Snow,Ice Pellets                                         0.8   
Thunderstorms                                           26.7   
Thunderstorms,Heavy Rain Showers                        10.9   
Thunderstorms,Moderate Rain Showers,Fog                 19.6   
Thunderstorms,Rain                                      21.3   
Thunderstorms,Rain Showers                              25.5   
Thunderstorms,Rain Showers,Fog                          22.9   
Thunderstorms,Rain,Fog                                  20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                                20.4         99   
Cloudy                                               22.6         99   
Drizzle                                              17.7         96   
Drizzle,Fog                                          19.1        100   
Drizzle,Ice Pellets,Fog                              -0.7         92   
Drizzle,Snow                                          0.2         95   
Drizzle,Snow,Fog                                      0.6         98   
Fog                                                  19.6       

# 

* If we want to find the maximum, minimum and mean of each weather factors on the rainy days, then we must specify it by using get_group() function as get_group("Rain).

In [9]:
data.groupby("Weather Condition").get_group("Rain").max(numeric_only=True)

Temp_degree celcius     22.80
Dew Point Temp_C        20.40
Rel Hum_%               99.00
Wind Speed_km/h         52.00
Visibility_km           48.30
Press_kPa              102.26
dtype: float64

In [10]:
data.groupby("Weather Condition").get_group("Rain").min(numeric_only=True)

Temp_degree celcius     0.30
Dew Point Temp_C       -5.70
Rel Hum_%              40.00
Wind Speed_km/h         0.00
Visibility_km           4.00
Press_kPa              97.52
dtype: float64

In [11]:
data.groupby("Weather Condition").get_group("Rain").mean(numeric_only=True)

Temp_degree celcius      9.786275
Dew Point Temp_C         7.042810
Rel Hum_%               83.624183
Wind Speed_km/h         19.254902
Visibility_km           18.856536
Press_kPa              100.233333
dtype: float64

# Q.13. Show all the records where the "Weather Condition" is "Fog".

In [12]:
data.groupby("Weather Condition").get_group("Fog")

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0        1/1/2012 0:00                 -1.8              -3.9         86   
1        1/1/2012 1:00                 -1.8              -3.7         87   
4        1/1/2012 4:00                 -1.5              -3.3         88   
5        1/1/2012 5:00                 -1.4              -3.3         87   
6        1/1/2012 6:00                 -1.5              -3.1         89   
...                ...                  ...               ...        ...   
8716   12/29/2012 4:00                -16.0             -17.2         90   
8717   12/29/2012 5:00                -14.8             -15.9         91   
8718   12/29/2012 6:00                -13.8             -15.3         88   
8719   12/29/2012 7:00                -14.8             -16.4         88   
8722  12/29/2012 10:00                -12.0             -13.3         90   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
0                   4            8.0     101.24               Fog  
1                   4            8.0     101.24               Fog  
4                   7            4.8     101.23               Fog  
5                   9            6.4     101.27               Fog  
6                   7            6.4     101.29               Fog  
...               ...            ...        ...               ...  
8716                6            9.7     101.25               Fog  
8717                4            6.4     101.25               Fog  
8718                4            9.7     101.25               Fog  
8719                7            8.0     101.22               Fog  
8722                7            6.4     101.15               Fog  

[150 rows x 8 columns]

In [17]:
data[data["Weather Condition"] == "Fog"]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
0        1/1/2012 0:00                 -1.8              -3.9         86   
1        1/1/2012 1:00                 -1.8              -3.7         87   
4        1/1/2012 4:00                 -1.5              -3.3         88   
5        1/1/2012 5:00                 -1.4              -3.3         87   
6        1/1/2012 6:00                 -1.5              -3.1         89   
...                ...                  ...               ...        ...   
8716   12/29/2012 4:00                -16.0             -17.2         90   
8717   12/29/2012 5:00                -14.8             -15.9         91   
8718   12/29/2012 6:00                -13.8             -15.3         88   
8719   12/29/2012 7:00                -14.8             -16.4         88   
8722  12/29/2012 10:00                -12.0             -13.3         90   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
0                   4            8.0     101.24               Fog  
1                   4            8.0     101.24               Fog  
4                   7            4.8     101.23               Fog  
5                   9            6.4     101.27               Fog  
6                   7            6.4     101.29               Fog  
...               ...            ...        ...               ...  
8716                6            9.7     101.25               Fog  
8717                4            6.4     101.25               Fog  
8718                4            9.7     101.25               Fog  
8719                7            8.0     101.22               Fog  
8722                7            6.4     101.15               Fog  

[150 rows x 8 columns]

# Q.14. Find all instances when weather is "Clear" or visibility is above 40.

In [18]:
data[(data["Weather Condition"] == "Clear") | (data["Visibility_km"] > 40)]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
67      1/3/2012 19:00                -16.9             -24.8         50   
106     1/5/2012 10:00                 -6.0             -10.0         73   
107     1/5/2012 11:00                 -5.6             -10.2         70   
108     1/5/2012 12:00                 -4.7              -9.6         69   
109     1/5/2012 13:00                 -4.4              -9.7         66   
...                ...                  ...               ...        ...   
8749  12/30/2012 13:00                -12.4             -16.2         73   
8750  12/30/2012 14:00                -11.8             -16.1         70   
8751  12/30/2012 15:00                -11.3             -15.6         70   
8752  12/30/2012 16:00                -11.4             -15.5         72   
8756  12/30/2012 20:00                -13.8             -16.5         80   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
67                 24           25.0     101.74             Clear  
106                17           48.3     100.45      Mainly Clear  
107                22           48.3     100.41      Mainly Clear  
108                20           48.3     100.38      Mainly Clear  
109                26           48.3     100.40      Mainly Clear  
...               ...            ...        ...               ...  
8749               37           48.3     100.92     Mostly Cloudy  
8750               37           48.3     100.96      Mainly Clear  
8751               32           48.3     101.05      Mainly Clear  
8752               26           48.3     101.15      Mainly Clear  
8756               24           25.0     101.52             Clear  

[3027 rows x 8 columns]

In [19]:
# Answer = 3027 times

* IMPORTANT: For mathematical operators like and,[ & ] and or[ | ], we cannot use groupby( ) and get_group( ) functions.

# Q.15. Find all instances when:

A. 'Weather Condition' is 'Clear' and 'Rel Hum_%' is greater than 50.

or, 

B. 'Visibility_km' is above 40.

* (A)

In [9]:
data[(data["Weather Condition"] == "Clear") & (data["Rel Hum_%"] > 50)]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
114     1/5/2012 18:00                 -7.1             -14.4         56   
115     1/5/2012 19:00                 -9.2             -15.4         61   
116     1/5/2012 20:00                 -9.8             -15.7         62   
117     1/5/2012 21:00                 -9.0             -14.8         63   
241     1/11/2012 1:00                -10.7             -17.8         56   
...                ...                  ...               ...        ...   
8646   12/26/2012 6:00                -13.4             -14.8         89   
8698  12/28/2012 10:00                 -6.1              -8.6         82   
8713   12/29/2012 1:00                -11.9             -13.6         87   
8714   12/29/2012 2:00                -11.8             -13.1         90   
8756  12/30/2012 20:00                -13.8             -16.5         80   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
114                11           25.0     100.71             Clear  
115                 7           25.0     100.80             Clear  
116                 9           25.0     100.83             Clear  
117                13           25.0     100.83             Clear  
241                17           25.0     101.49             Clear  
...               ...            ...        ...               ...  
8646                4           25.0     102.47             Clear  
8698               19           24.1     101.27             Clear  
8713               11           25.0     101.31             Clear  
8714               13           25.0     101.33             Clear  
8756               24           25.0     101.52             Clear  

[1070 rows x 8 columns]

In [10]:
# Answer: 1070 times

* (B)

In [11]:
data[data["Visibility_km"] > 40]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
106     1/5/2012 10:00                 -6.0             -10.0         73   
107     1/5/2012 11:00                 -5.6             -10.2         70   
108     1/5/2012 12:00                 -4.7              -9.6         69   
109     1/5/2012 13:00                 -4.4              -9.7         66   
110     1/5/2012 14:00                 -5.1             -10.7         65   
...                ...                  ...               ...        ...   
8748  12/30/2012 12:00                -12.2             -15.7         75   
8749  12/30/2012 13:00                -12.4             -16.2         73   
8750  12/30/2012 14:00                -11.8             -16.1         70   
8751  12/30/2012 15:00                -11.3             -15.6         70   
8752  12/30/2012 16:00                -11.4             -15.5         72   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
106                17           48.3     100.45      Mainly Clear  
107                22           48.3     100.41      Mainly Clear  
108                20           48.3     100.38      Mainly Clear  
109                26           48.3     100.40      Mainly Clear  
110                22           48.3     100.46      Mainly Clear  
...               ...            ...        ...               ...  
8748               26           48.3     100.91     Mostly Cloudy  
8749               37           48.3     100.92     Mostly Cloudy  
8750               37           48.3     100.96      Mainly Clear  
8751               32           48.3     101.05      Mainly Clear  
8752               26           48.3     101.15      Mainly Clear  

[2014 rows x 8 columns]

In [12]:
# Answer: 2014 times

* (A and B combined)

In [24]:
data[((data['Weather Condition'] == 'Clear') & (data["Rel Hum_%"] > 50)) | (data["Visibility_km"] > 40)]

Date/Time  Temp_degree celcius  Dew Point Temp_C  Rel Hum_%  \
106     1/5/2012 10:00                 -6.0             -10.0         73   
107     1/5/2012 11:00                 -5.6             -10.2         70   
108     1/5/2012 12:00                 -4.7              -9.6         69   
109     1/5/2012 13:00                 -4.4              -9.7         66   
110     1/5/2012 14:00                 -5.1             -10.7         65   
...                ...                  ...               ...        ...   
8749  12/30/2012 13:00                -12.4             -16.2         73   
8750  12/30/2012 14:00                -11.8             -16.1         70   
8751  12/30/2012 15:00                -11.3             -15.6         70   
8752  12/30/2012 16:00                -11.4             -15.5         72   
8756  12/30/2012 20:00                -13.8             -16.5         80   

      Wind Speed_km/h  Visibility_km  Press_kPa Weather Condition  
106                17           48.3     100.45      Mainly Clear  
107                22           48.3     100.41      Mainly Clear  
108                20           48.3     100.38      Mainly Clear  
109                26           48.3     100.40      Mainly Clear  
110                22           48.3     100.46      Mainly Clear  
...               ...            ...        ...               ...  
8749               37           48.3     100.92     Mostly Cloudy  
8750               37           48.3     100.96      Mainly Clear  
8751               32           48.3     101.05      Mainly Clear  
8752               26           48.3     101.15      Mainly Clear  
8756               24           25.0     101.52             Clear  

[2921 rows x 8 columns]

In [22]:
# Answer = 2921 times